### ISSUE TITLE
Dynamic bundle and content loading

### ISSUE THREAD

**Issue by mfellner** [zalando-incubator/tessellate#34]
tessellate-fragment should dynamically load bundles and content (properties) from configurable sources.

- [x] retrieve unified response configuration values (determine bundle and property source) from
    * the request headers
    * request URL parameters
    * environment configuration
- [x] additional, overriding properties are loaded from an optionally provided property source

**Comment by semonte** [comment]
I think we need to also consider a case where the content endpoint is OAuth 2 protected.

If no authentication is needed, the following config would be enough:

```
{
	bundle_url: "https://...",
	content_url: "https://...",
}
```

If authenticated requests are needed for content endpoint, we need to provide information about the OAuth 2 authentication. As an example:

```
{
  "bundle_url": "",
  "content_url": "",
  "accessTokenUri": "",
  "scopes": []
}
```

Thoughts?

**Comment by mfellner** [comment]
Good point. The bundle endpoint must be public in order to load the JavaScript in the browser as well but the content endpoint may be protected. I would suggest the following schema for the complete application configuration:

```yaml
---
# ... other application configuration
sources:
  bundles:
    src: https://cdn.com/bundles  # base address
    path: /my/bundle              # optional - /index.js, /index.css etc. are expected to exist
    query:                        # optional query params
      key: value
  properties:  # i.e. content - 'properties' indicates the content must work as React props
    src: https://server.com/content
    path: /some/path
    query:
      key: value
    auth:                         # optional authentication information
      type: oauth2
      accessTokenUri: https://secure.com/oauth2/access_token?realm=/services
      scopes:
        - content.read
```

The configuration for `sources` would be retrieved from the following hierarchy of sources:

1. nconf (environment variables, application parameters, config file, etc.)
2. Remote configuration file: the location of the file would be encoded in the `sources` query parameter. Example fragment tag:
```
<z-fragment src="https://tessellate-fragment.zalan.do/fragment?sources=https%3A%2F%2Fcdn.com%2Fsources.json"></z-fragment>
```
3. A function that receives HTTP request headers and URL parameters and returns a configuration object (not sure how to make that externally configurable):
```javascript
(headers: {[key: string]: string}, params: {[key: string]: string}) => {[key: string]: Object}
```

This way some configuration, like the base addresses, can come from nconf while the rest can be different for each request (based the request information).

**Comment by mfellner** [comment]
Here's an idea for supporting header- oder query param values:

```yaml
sources:
  bundles:
    src: https://cdn.com/bundles    # base address
    path:                           # string | Object
      type: header                  # 'header' | 'query'
      key: x-request-uri            # header or query param name
      regexp: :\/\/(.+\/[^\/?&]+)   # e.g. new RegExp(...).exec(headers[...])[1]
```

The configuration `sources.bundles.path` can be a literal string or an object with type (header or query), key and optional regexp (for parsing the value).

For full flexibility, any value, e.g. `sources.bundles.src`, `sources.bundles.path`, `sources.properties.path` etc., should support this.

I would suggest adding this as another improvement later on and keep things static in the beginning.

**Comment by semonte** [comment]
Good idea, and I also agree that let's implement this as an improvement later on.

**Comment by semonte** [comment]
This can be closed I suppose, features were added in #35 and #45.
